File: Algorithm/back_tracing/Subsets/Subsets.cpp
class Solution {
	public:
		vector<vector<int>> subsets(vector<int>& nums) {
			int len = nums.size();
			int start = 0;
			vector<vector<int>> list;
			vector<int> templist;
			traceback(start, len, list, templist, nums);
			return list;
		}

		void traceback(int start, int len, vector<vector<int>>&list, vector<int> templist, vector<int>&nums)
		{
			list.push_back(templist);
			for (int i = start; i<len; i++)
			{
				templist.push_back(nums[i]);
				traceback(++i, len, list, templist, nums);
				templist.pop_back();
			}
		}
	};

File: Algorithm/back_tracing/Subsets/subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        lst = []
        n = len(nums)

        def helper(i,tmplst):
            lst.append(tmplst)
            for j in range(i,n):
                helper(j+1,tmplst+[nums[j]])

        helper(0,[])
        return lst
